idl/storage_service.idl.hh

```diff
@@ -29,7 +29,8 @@ struct raft_topology_cmd {
         barrier,
```

Comment:
> We dont't treat wait_for_ip errors as fatal, we just write a warning into the log if the command failed. The gossiper events might have been delayed for various reasons,

That's true but isn't that precisely why we wait? The gossiper event with IP must reach us eventually. Did you observe in tests that it never arrives?

Reply:
Ah, I think I get it -- this is for the case of bootstrap failing, therefore the gossiper event may never reach us. We might not be able to contact the node that failed to boot.

But the comment should elaborate on why the try/catch is needed.

In fact, why not put the `wait_for_ip` from beginning of `case node_state::none` into the case we want to reject the node? The case where we don't reject should be handled by the global barrier below, IIUC.

Reply:
I also don't understand why we can continue if wait_for_ip fails. We don't have the IP, we cannot communicate with the node -- how can we proceed?

Reply:
> Ah, I think I get it -- this is for the case of bootstrap failing, therefore the gossiper event may never reach us. We might not be able to contact the node that failed to boot.

Yes, but not only. The main reason is in the commit message:
> the missing IP for some node shouldn't be fatal for data plane
> request, this node should be treated just like any other
> unavailable node.

> In fact, why not put the wait_for_ip from beginning of case node_state::none into the case we want to reject the node? The case where we don't reject should be handled by the global barrier below, IIUC.

Why? Now the code is more straightforward for me. But we need to add `get_my_id` to the `exclude_nodes`...

Reply:
I don't understand what data plane has anything to do with the issue. This is control plane code.

Reply:
Does this control plane code have to communicate with the node?
If yes, we need the IP. I don't understand how we can skip getting the IP if we need to communicate with it.
If no, then we don't need the wait_for_ip at all.

Reply:
I understand that this part
```cpp
                        node.guard = co_await exec_global_command(std::move(node.guard),
                            raft_topology_cmd::command::wait_for_ip,
                            {node.id});
```

is necessary for the data plane to function.

But I also don't understand why you made it optional. We should fail bootstrap if we cannot learn the bootstrapping node's IP.

Reply:
It is a philosophical question. If you proceed here and the IP will not be resolved before it is needed you will fail later. You already have (or suppose to have) code that handles the failure, so you may as well proceed here.

Reply:
> I don't understand what data plane has anything to do with the issue. This is control plane code.

The moment where we absolutely need an IP is when a node is coordinating a data plane request and it needs to direct it to another node. This is where we want to treat a missing IP for a node the same way as the node just being unavailable. This is why we have `try/catch` around `wait_for_ip` cmd call.

Reply:
> Does this control plane code have to communicate with the node?
> If yes, we need the IP. I don't understand how we can skip getting the IP if we need to communicate with it further down the line.
> If no, then we don't need the wait_for_ip at all.

We need an IP to deliver `join_node_response_params::rejected` to the node. We need to `wait_for_ip` before `update_topology_state(state := left)` since otherwise the node will be banned and we'll never get an IP. We need try/catch around `wait_for_ip(node.id)` because we need `update_topology_state(state := left)` for it event if we can't resolve an IP.

Reply:
> But I also don't understand why you made it optional. We should fail bootstrap if we cannot learn the bootstrapping node's IP.

Why fail bootstrap if some of the nodes can't resolve new node's IP? The other nodes can still work with it. And even if node X can't resolve ip of node Y it still can coordinate requests, just treating Y as unavailable.

Reply:
> Why fail bootstrap if some of the nodes can't resolve new node's IP? The other nodes can still work with it. And even if node X can't resolve ip of node Y it still can coordinate requests, just treating Y as unavailable.

By making the `wait_for_ip` optional you're reducing data path availability. For example temporarily CL=ALL requests might be failing during bootstrap, because we proceed without waiting for all nodes to learn about the node's IP. This is not a correct thing to do. Topology changes should not be introducing unavailability if there are no failures.

And if they cannot learn the IP eventually, it means there's some larger problem. If they can't learn it now, during boot, when the node is alive and gossiping with the cluster, why would it be a good idea to proceed with bootstrap? Any reason they will be able to learn the IP of this node when it finishes bootstrapping, if they can't learn it now? You'll just reduce availability even further once the node becomes NORMAL. So it's logical to fail bootstrap in this case.

Reply:
Above I was talking about the global `wait_for_ip`, the purpose of which is to support the data plane.

The other `wait_for_ip`, at the beginning of the transition, is used to support the control plane. And there it makes sense to make it optional, because delivering `join_node_response_params::rejected` is optional. But it needs a comment. And because it's used for the singular purpose of delivering `join_node_response_params::rejected`, I think it would be clearer if the wait was inside the branch that delivers `join_node_response_params::rejected`.

Reply:
made it non-optional